sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala

```diff
@@ -121,6 +123,42 @@ class StatefulProcessorHandleImpl(
 
   override def getQueryInfo(): QueryInfo = currQueryInfo
 
+  private def getTimerState[T](): TimerStateImpl[T] = {
+    new TimerStateImpl[T](store, timeoutMode, keyEncoder)
+  }
+
+  private val timerState = getTimerState[Boolean]()
+
+  override def registerTimer(expiryTimestampMs: Long): Unit = {
+    verify(timeoutMode == ProcessingTime || timeoutMode == EventTime,
+    s"Cannot register timers with incorrect TimeoutMode")
+    verify(currState == INITIALIZED || currState == DATA_PROCESSED,
+    s"Cannot register timers with " +
+      s"expiryTimestampMs=$expiryTimestampMs in current state=$currState")
+
+    if (timerState.exists(expiryTimestampMs)) {
+      logWarning(s"Timer already exists for expiryTimestampMs=$expiryTimestampMs")
```

Comment:
We should validate here that `expiryTimestampMs` is >= `batchTimestampMs`, or watermark based on the timeout mode.

Reply:
I thought we decided not to do this ? So basically we allow for a timestamp earlier than latest to be generated which would fire the timer right away in the next invocation

Reply:
As we process timers after the data, this would result in firing of the timer right away in the same micro-batch after the data is processed. I think that is fine, just pointing it out. 

Also, equality should take care of firing the timer right away though, isn't it?

Reply:
Yes thats correct. Will fire within the same microbatch